exporter/internal/queue/sized_elements_channel.go

```diff
@@ -0,0 +1,113 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package queue // import "go.opentelemetry.io/collector/exporter/internal/queue"
+
+import "sync/atomic"
+
+// sizedElementsChannel is a channel wrapper for sized elements with a capacity set to a total size of all the elements.
+// The channel will accept elements until the total size of the elements reaches the capacity.
+type sizedElementsChannel[T any] struct {
+	used *atomic.Int64
+
+	// We need to store the capacity in a separate field because the capacity of the channel can be higher.
+	// It happens when we restore a persistent queue from a disk that is bigger than the pre-configured capacity.
+	cap int64
+	ch  chan T
+}
+
+type sizedElementsChannelOption[T any] func(*sizedElementsChannel[T])
+
+// withPreloadElements puts the elements into the queue with the given size. It's used by the persistent queue to
+// initialize the queue with the elements recovered from the disk.
+func withPreloadElements[T any](els []T, totalSize int64) sizedElementsChannelOption[T] {
```

Comment:
Since no other options for the moment, I would not make this an option but another ctor:

```golang
func newSizedElementsChannelWithPreloadElements[T any](capacity int64, els []T, totalSize int64) *sizedElementsChannel[T] {
```

Reply:
Makes sense, updated